File: docs/sort/code/insertion.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "example.c"

void sort(char* s) {
	for (int i = 1; i < strlen(s); i++) {
		for (int j = i; j > 0; j--) {
			if (less(s[j], s[j - 1])) {
				exch(s, j, j - 1);
			} else {
				break;
			}
		}
	}
}

int main() {
	char* c;

	printf("pls input array:\n");
	scanf("%s", c);

	printf("---Before---\n");
	show(c);

	sort(c);

	printf("---After---\n");
	show(c);

	return EXIT_SUCCESS;
}

File: docs/sort/code/insertion.ts
import {exch, less} from './example';

function insertionSort(arr: number[]) {
  // 遍历未排区间
  for (let i = 1; i < arr.length; i++) {
		// 反向遍历已排区间
		// 将 a[i] 插入到 a[i-1]、a[i-2]、a[i-3]...之中
    for (let j = i; j > 0 && less(arr[j], arr[j - 1]); j--) {
      exch(arr, j, j - 1);
    }
  }
}
